flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvReaderFormat.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.formats.csv;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.connector.file.src.reader.SimpleStreamFormat;
+import org.apache.flink.connector.file.src.reader.StreamFormat;
+import org.apache.flink.core.fs.FSDataInputStream;
+import org.apache.flink.formats.common.Converter;
+
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.MappingIterator;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.dataformat.csv.CsvMapper;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.dataformat.csv.CsvSchema;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * A {@code StreamFormat} for reading CSV files.
+ *
+ * <p>The following example shows how to create a {@code CsvReaderFormat} where the schema for CSV
+ * parsing is automatically derived based on the fields of a POJO class.
+ *
+ * <pre>{@code
+ * CsvReaderFormat<SomePojo> csvFormat = CsvReaderFormat.forPojo(SomePojo.class);
+ * FileSource<SomePojo> source =
+ *         FileSource.forRecordStreamFormat(csvFormat, Path.fromLocalFile(filesPath)).build();
+ * }</pre>
+ *
+ * <i> Note: you might need to add {@code @JsonPropertyOrder({field1, field2, ...})} annotation from
+ * the {@code Jackson} library to your class definition with the fields order exactly matching those
+ * of the CSV file columns).</i>
+ *
+ * <p>If you need more fine-grained control over the CSV schema or the parsing options, use the more
+ * low-level {@code forSchema} static factory method based on the {@code Jackson} library utilities:
+ *
+ * <pre>{@code
+ * CsvReaderFormat<T> forSchema(CsvMapper mapper,
```

Comment:
Nit: The example does not look very helpful. Maybe you can show a code snipped in conjunction with using the format with the FileSource and it would also be great two a second example in the same code block showing it for a POJO class.